### ISSUE TITLE
Configuration

### ISSUE THREAD

**Issue by romanofski** [purebred-mua/purebred#8]
This touches a little bit on the architecture as well, but I wonder how users should be able to configure the tool. Do we use a high level language like YAML or <whatever> to make it easy to configure? There are some benefits from using e.g. Haskell as well, although it is less portable in the sense if you'd use configuration management you wouldn't easily able to template it.

Being an XMonad user, there seems to be quite a big benefit at how XMonad is architected. Your config which you compile pulls in core and plugins. That is very flexible from writing plugins by a community, although not very end-user friendly who don't want to get into the Haskell details...

I would leaning almost not to put too much weight on end-user friendliness, but think of it as a professional who would be smart enough to pick up some Haskell to configure his tool.

**Comment by frasertweedale** [comment]
The user base will be folks who already lean towards mutt and similar.  I would not worry too
much about what configuration looks like, and go for whatever approach (a) gives us the features
we need and (b) gives users the power we want them to have (and no more).

**Comment by romanofski** [comment]
I suppose one thing is certain and that is that we need some sort of configuration for a first version.

Thinking about it, I suppose there would always be a chance to take the XMonad route later, although it'll be costly. The benefits of having a more pluggable architecture is not really depended on that approach tho, so providing way for third party developers to write plugins is still possible.

So maybe: go with configurator for a 1.0 release, but create an issue for creating a pluggable architecture?

**Comment by romanofski** [comment]
I found https://hackage.haskell.org/package/configurator-ng and in the manifesto this:

> The application interface of configurator has numerous problems: 
>     * it makes it easy to introduce race conditions
>     * it makes it difficult to write an application that is relatively robust to misconfiguration errors
>     * does not scale well to moderately complex configuration scenarios
>     * the configuration change notifications are particularly difficult to use beyond the most trivial of use cases.

Given that we don't really need the ability of automatically reloading configuration and taking a bit of a fancy from https://github.com/matterhorn-chat/matterhorn maybe we should just go with a normal ini parser?

**Comment by frasertweedale** [comment]
On Mon, Jul 24, 2017 at 12:53:13AM +0000, Roman Joost wrote:
> Given that we don't really need the ability of automatically
> reloading configuration and taking a bit of a fancy from
> https://github.com/matterhorn-chat/matterhorn maybe we should just
> go with a normal ini parser?
> 
Just pick something *simple* now so it is easy to change later if we
want.

**Comment by romanofski** [comment]
Turns out following the XMonad approach is actually easier than I thought. I've looked it up:

https://github.com/xmonad/xmonad/blob/master/src/XMonad/Main.hs#L114

**Comment by frasertweedale** [comment]
On Wed, Aug 09, 2017 at 04:07:07AM +0000, Roman Joost wrote:
> Turns out following the XMonad approach is actually easier than I thought. I've looked it up:
> 
> https://github.com/xmonad/xmonad/blob/master/src/XMonad/Main.hs#L114
> 
We may not even need to compile the config, if the `runhaskell` or
`runghc` program is available we can just run the program as a
script.  But I'm not sure about overhead etc.

**Comment by frasertweedale** [comment]
Alternatively we use ghc-as-a-library or invoke ghc itself to compile the config if we
detect that the compiled version is not available, or if the last modified time of the
"source" is newer than the timestamp of the compiled binary.

IMO running it as a script is sufficient until proven otherwise.

**Comment by romanofski** [comment]
Fixing #31 made the app/Main.hs a  one liner:

```Haskell
main = void $ M.defaultMain theApp =<< initialState =<< getDatabasePath
```
I try to keep it that way using a default configuration (or I guess with minimal difference with a custom config)

**Comment by frasertweedale** [comment]
noice.

hey look at this: the xmonad `recompile` function that recompiles the user's "config"
when it is newer than the compiled lib, or when forced.  For your inspiration :)

**Comment by romanofski** [comment]
So this is basically done. I'm not sure how well this works, since our API at this point changes a lot. Maybe you want to give it a try and see how you go?

**Comment by romanofski** [comment]
Pushed it to master now. Give it a go if you like. I'll close this and create new items for smaller bits and pieces.

**Comment by romanofski** [comment]
Reopening this, since at this point it's not possible to run the configuration by importing just purebred.

**Comment by romanofski** [comment]
I've now spent some time studying the XMonad configuration and put it into practice in Purebred. I've needed a bunch of re-exports of our dependencies, but got it working (see b500a199d5c21b78eb04c3c0a8ddcc5436f0989b). I've extended it a bit, so that it is possible to overwrite the default config directory and to use the GHC provided by stack. That is mainly for testing purposes, since I'm not going to copy around files all the time.

Overall, due to sucky key binding implementation it still is quite cumbersome to overwrite/configure different key bindings. I'm sure there are other parts which are not configurable yet as well. I think the more we start using it, the more we'll see spots which need refactoring in order to make this easier ...

**Comment by romanofski** [comment]
I'm a bit indecisive how to move forward. The default configuration includes performing an IO action to set the notmuch database path. Furthermore, the main application also accepted to explicitly provide a database path which should have the preference.

If I divorce the configuration from performing an additional IO action to get the notmuch database path or expecting an additional parameter (which sounds sensible), it still means something/someone needs to know to set it somehow before we start the application.

Possible solution: The purebred command which is the new entry point (see https://github.com/purebred-mua/purebred/commit/b500a199d5c21b78eb04c3c0a8ddcc5436f0989b#diff-54a51add1191c97b14ccda5e4dc2d6f7R63), could have some sort of stitch-up command which does the command line parsing and either performs the IO action to get the notmuch database path or sets the explicitly set database provided by the command line. Downside is, that the user won't be able to set the database path.

@frasertweedale Any comments for this?

**Comment by frasertweedale** [comment]
@romanofski let me rephrase to see if I understand correctly:

1. You want to provide DB path in configuration (with a sensible default).
2. But to build the configuration, you need to do stuff in `IO` (for the default, anyway).
3. Ergo, you can't supply the configuration as a pure value.  Problem.

Did I accurately interpret the problem statement?

If so, here's my $0.02 on how to work this:

Instead of `NotmuchSettings` having `_nmDatabase :: FilePath`, make it `_nmDatabase :: IO FilePath`.  That way, the default configuration can be the *computation* that, when run, yields the database
path.  Users wanting to override it with a pure value can set it to `(pure "/path/to/notmuch-db")`.

When the program gets run, it would be desirable desirable to run the computation immediately
so that the pure value is known throughout the rest of the program.  So to avoid completely
duplicating the whole `Configuration` and `NotmuchSettings` data structures, abstract over just
that field.  So you end up with something like:

```haskell
data NotmuchSettings a =
 { _nmSearch :: String
 , _nmDatabase :: a
 , _nmNewTag = String
 }
```
And the `nmDatabase` optic becomes:
```haskell
nmDatabase :: Lens (NotmuchSettings a) (NotmuchSettings b) a b
nmDatabase f (NotmuchSettings a b c) = fmap (\b' -> NotmuchSettings a b' c) (f b)
```

The type variable gets propagated up into the `Configuration` data type as well.

```haskell
data Configuration a = Configuration
    { _confColorMap :: Brick.AttrMap
    , _confNotmuch :: NotmuchSettings a   -- along with an update to the optic, of course
    , _confEditor :: T.Text
    , _confMailView :: MailViewSettings
    , _confIndexView :: IndexViewSettings
    , _confComposeView :: ComposeViewSettings
}
```

A couple of type synonyms can delineate between the configuration type the user
provides, and the "processed" configuration type internally within the program:

```haskell
type UserConfiguration = Configuration (IO FilePath)
type InternalConfiguration = Configuration (FilePath)
```

Using this approach will allow us to add more type variables to deal with any other
discrepancies between *what user gives us* and *what we want to work with inside
the program*, if/when other such discrepancies arise.

Then you just need a function:
```haskell
processUserConfiguration :: UserConfiguration -> IO InternalConfiguration
processUserConfiguration = _anExerciseForTheImplementor
```

What do you think?

**Comment by frasertweedale** [comment]
@romanofski is there a ref that contains b500a199d5c21b78eb04c3c0a8ddcc5436f0989b ?

If not could you please push a branch or something containing it (`master` is fine if it's stable),
and I will do the changes discussed above.

**Comment by romanofski** [comment]
On August 30, 2017 7:31:04 PM GMT+10:00, Fraser Tweedale <notifications@github.com> wrote:
>@romanofski is there a ref that contains
>b500a199d5c21b78eb04c3c0a8ddcc5436f0989b ?
>
>If not could you please push a branch or something containing it
>(`master` is fine if it's stable),
>and I will do the changes discussed above.

It should be my xmonad_config branch on my fork. If you can't find it let me know and I'll push it to the main repo.
-- 
Sent from a mobile device.

**Comment by frasertweedale** [comment]
@romanofski no worries, I've found it there, cheers.

**Comment by frasertweedale** [comment]
@romanofski done; please review branch https://github.com/purebred-mua/purebred/tree/xmonad_config

This was about a 15 minute job :)

The rest of the time was spent wrangling with the new program launching
paradigm and stack >_<

**Comment by romanofski** [comment]
W00t cheers. I gave up last night being too tired. Seems like I was on the right path, but should have studied over instead of trying to do everything with view/set.

**Comment by romanofski** [comment]
Alright, pushed to master. I think I can finally close this.

The test for this is in a separate backlog item. I've got to say that having an easily installable OS package would make it much easier to configure, just because you'd be using the global GHC with it's dependencies. I would have already created one, but atm Copr is broken for Haskell builds. Until it's fixed I think I just keep going ...
